### PR TITLE
Sugar improvements

### PR DESCRIPTION
**Pull request by konsumlamm**
* use `runnableExamples`
* remove unnecessary import (`typetraits`)

### CODE REVIEW

File: lib/pure/sugar.nim [hunk 1]
```diff
@@ -344,51 +337,55 @@ macro collect*(init, body: untyped): untyped {.since: (1, 1).} =
...
-    assert x == @["bird"]
     ## HashSet:
-    let y = initHashSet.collect:
+    let y = collect(initHashSet()):
       for d in data.items: {d}
+    doAssert y == data.toHashSet
 
-    assert y == data.toHashSet
     ## Table:
     let z = collect(initTable(2)):
```
Comment (timotheecour) [inline lib/pure/sugar.nim:362]: `initOrderedTable` so that the test is robust against table implementation changes

Comment (timotheecour) [reply lib/pure/sugar.nim:362]: nevermind, `doAssert z == {0: "bird", 1: "word"}.toTable` makes it robust

File: lib/pure/sugar.nim [hunk 2]
```diff
@@ -193,21 +185,24 @@ proc freshIdentNodes(ast: NimNode): NimNode =
...
+        if i * j == 42:
+          capture i, j:
+            myClosure = () => fmt"{i} * {j} = 42"
+    doAssert myClosure() == "6 * 7 = 42"
+
+    let m = @[(s: string) => "to " & s,
+              (s: string) => "not to " & s]
+    let l = m.mapIt(capture(it, (s: string) => it(s)))
+    let r = l.mapIt(it("be"))
+    doAssert fmt"{r[0]}, or {r[1]}" == "to be, or not to be"
```
Comment (timotheecour) [inline lib/pure/sugar.nim:204]: pre-existing: this example is silly/bad, feel free to simplify as you see fit

**Review by timotheecour — APPROVED** [review]

File: lib/pure/sugar.nim [hunk 3]
```diff
@@ -160,16 +156,12 @@ macro dump*(x: untyped): untyped =
...
-  ##     x = 10
-  ##     y = 20
-  ##   dump(x + y)
-  ##
-  ## will print ``x + y = 30``.
+  runnableExamples:
+    let
+      x = 10
+      y = 20
+    dump(x + y) # will print `x + y = 30`
```
Comment (ringabout) [inline lib/pure/sugar.nim:163]: please do not print strings here.

File: lib/pure/sugar.nim [hunk 4]
```diff
@@ -11,7 +11,7 @@
 ## macro system.
 
 import std/private/since
-import macros, typetraits
+import macros
```
Comment (ringabout) [inline lib/pure/sugar.nim]: nit: use `import std/macros`.

File: lib/pure/sugar.nim (hunk 3 above)
Comment (timotheecour) [inline lib/pure/sugar.nim:163]:
```suggestion
    if false: dump(x + y) # if true would print `x + y = 30`
```

Comment (timotheecour) [reply lib/pure/sugar.nim:163]: in future PR we should deprecate `dump` in favor of a proc that returns a string; more generally, API's should almost never `echo`, that should be left up to user code (it's always less flexible, can't as easily be unit-tested, etc)

Comment (Araq) [reply lib/pure/sugar.nim:163]: Returning a string produces evil allocations though... And `dup` on a `var string` is too inconvenient for `dump`. *shrug*

File: lib/pure/sugar.nim (hunk 2 above)
Comment (konsumlamm) [reply lib/pure/sugar.nim:204]: I just realized the example also works without the `capture`, so we should definitely change it.

File: lib/pure/sugar.nim (hunk 3 above)
Comment (Clyybber) [reply lib/pure/sugar.nim:163]: Why? Examples should be simple, not unneccessarily convoluted/wrapped.

File: lib/pure/sugar.nim [hunk 5]
```diff
@@ -160,16 +156,12 @@ macro dump*(x: untyped): untyped =
...
-  ##     x = 10
-  ##     y = 20
-  ##   dump(x + y)
-  ##
-  ## will print ``x + y = 30``.
+  runnableExamples:
+    let
+      x = 10
+      y = 20
+    if false: dump(x + y) # would print `x + y = 30`
```
Comment (Clyybber) [inline lib/pure/sugar.nim:163]:
```suggestion
    dump(x + y) # would print `x + y = 30`
```

File: lib/pure/sugar.nim [hunk 6]
```diff
@@ -160,16 +156,12 @@ macro dump*(x: untyped): untyped =
...
-  ##     x = 10
-  ##     y = 20
-  ##   dump(x + y)
-  ##
-  ## will print ``x + y = 30``.
+  runnableExamples:
+    let
+      x = 10
+      y = 20
+    dump(x + y) # would print `x + y = 30`
```
Comment (Clyybber) [inline lib/pure/sugar.nim:163]:
```suggestion
    dump(x + y) # will print `x + y = 30`
```

File: lib/pure/sugar.nim (hunk 2 above)
Comment (konsumlamm) [reply lib/pure/sugar.nim:204]: See #16820.

File: lib/pure/sugar.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/sugar.nim:163]: > Returning a string produces evil allocations though

I don't see how that's a problem for a proc that's meant for debugging templates/macros

> And dup on a var string is too inconvenient for dump. shrug

no need for dup or var string. see https://github.com/nim-lang/Nim/pull/16841

Comment (timotheecour) [reply lib/pure/sugar.nim:163]: > Why?

https://github.com/nim-lang/Nim/pull/16802#discussion_r563438501

**Pull request merged**